### PR TITLE
Rewrite input management

### PR DESCRIPTION
**Pull request by paked**
This is the result of #252.

I'd really appreciate some code review while I am finishing everything off.

**TODO**

- [x] Handle different button states apart from justPressed and justReleased
- [x] Add option for defaults
- [x] Fix keyboardScroller
- [x] Check if I need to mutex-ify
- [x] Fix demos
- [x] Documentation

### CODE REVIEW

File: axis.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package engo
+
+type Axis struct {
```
Comment (otraore) [inline axis.go]: Try to document all exported things

File: axis.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package engo
+
+// Axis is an input which is a spectrum of values. An example of this is the horizontal movement in a game, or how far a joystick is pressed.
```
Comment (otraore) [inline axis.go]: Nice work on documentation

File: button.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package engo
+
+// Button is an input which can be either JustPressed, JustReleased or Down. Common uses would be for, a jump key or an action key.
+type Button struct {
+	Triggers []Key
+	Name     string
+}
+
+func (b Button) JustPressed() bool {
```
Comment (otraore) [inline button.go]: It'd be nice if you documentation JustPressed, JustReleased, Down

Comment (paked) [inline button.go]: I thought the cases here were pretty self explanatory...

Comment (otraore) [inline button.go]: They are but It's still nice for both godoc and the function docs that show up on editors.

Comment (paked) [inline button.go]: done

**Comment by paked** [comment]
@otraore @EtienneBruines Feel free to squash and merge. Will get mouse and joystick support in future respective PRs.

**Comment by mewmew** [comment]
@paked Great work or the new input system. Thanks for pushing this!

File: button.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
+package engo
+
+// Button is an input which can be either JustPressed, JustReleased or Down. Common uses would be for, a jump key or an action key.
+type Button struct {
+	Triggers []Key
+	Name     string
+}
+
+// JustPressed checks whether an input was pressed in the previous frame>
```
Comment (mewmew) [inline button.go]:
```diff
-frame>
+frame.
```

Comment (paked) [inline button.go]: fixed

**Comment by paked** [comment]
@otraore @EtienneBruines I've fixed gopherjs compilation, ready to merge!

**Comment by otraore** [comment]
LGTM

**Pull request merged**